docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,192 @@
+NumPy’s API was developed years before static type checking became part of the Python language, and follows Python’s historic recommendations to use a [duck-typing](https://docs.python.org/3/glossary.html#term-duck-typing)/[EAFP](https://docs.python.org/3/glossary.html#term-eafp) coding style, in which strict type-checking at runtime is discouraged. As a concrete example of this, consider the {func}`numpy.tile` function, which is defined like this:
+
+```python
+def tile(A, reps):
+  try:
+    tup = tuple(reps)
+  except TypeError:
+    tup = (reps,)
+  d = len(tup)
+  ...
+```
+
+Here the *intent* is that `reps` would contain either an `int` or a sequence of `int` values, but the *implementation* allows `tup` to be any iterable.
+When adding annotations to this kind of duck-typed code, we could take one of two routes:
+1. We may choose to annotate the *intent* of the function's API, which here might be something like `reps: Union[int, Sequence[int]]`.
+2. Conversely, we may choose to annotate the *implementation* of the function, which here might look something like `reps: Union[ConvertibleToInt, Iterable[ConvertibleToInt]]` where `ConvertibleToInt` is a special protocol that covers the exact mechanism by which our function converts the inputs to integers (i.e. via `__int__`, via `__index__`, via `__array__`, etc.). Note also here that in a strict sense, `Iterable` is not sufficient here because there are objects in Python that duck-type as iterables but do not satisfy a static type check against `Iterable` (namely, an object that is iterable via `__getitem__` rather than `__iter__`.)
+
+The advantage of #1, annotating intent, is that the annotations are more useful to the user in communicating the API contract; while for the developer the flexibility leaves room for refactoring when necessary. The down-side (particularly for gradually-typed APIs like JAX's) is that it's quite likely that user code exists which runs correctly, but would be flagged as incorrect by a type checker.
+Gradual typing of an existing duck-typed API means that the current annotation is implicitly `Any`, so changing this to a stricter type may present to users as a breaking change.
+
+Broadly speaking, annotating intent better serves Level 1 type checking, while annotating implementation better serves Level 3, while Level 2 is more of a mixed bag (both intent and implementation are important when it comes to annotations in IDEs).
+
+## JAX type annotation roadmap
+
+With this framing (Level 1/2/3) and JAX-specific challenges in mind, we can begin to develop our roadmap for implementing consistent type annotations across the JAX project.
+For JAX type annotation, we will be guided by the following principles:
+
+### Purpose of type annotations
+
+We would like to support full, *Level 1, 2, and 3* type annotation as far as possible. In particular, this means that we should have restrictive type annotations on both inputs and outputs to public API functions.
```

Comment:
> to support full, *Level 1, 2, and 3* type annotation as far as possible

From reading the rest of this note, and the suggestive ordering of those numbers, it seems like you have some degree of prioritization/preference in mind (e.g. 1 typically more important than 3). Is that the case?

Reply:
It's not an order of preference, it's an order of how complicated/specific the types must be.

Reply:
(my own opinion is that Level 3 is usually more trouble than its worth, especially for gradual typing of established packages where APIs have developed in a free-form duck-typing world typical of historical Python practices, but our loudest users seem pretty intent on Level 3 types wherever possible)

Reply:
Nice – that's clarifying. I wasn't sure what the level numbers were meant to convey exactly.

> my own opinion is that comprehensive Level 3 is usually more trouble than its worth [...] but [...] users seem pretty intent on Level 3 types wherever possible

Maybe state this? We want to have concrete choices to act on (as you say, this is a "roadmap"). I'd be supportive of something like "we should at least aim for 2, though note some want 3 and that can come next" – or whatever you'd actually propose instead. We're not going to have a fully annotated codebase tomorrow.